services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -0,0 +1,122 @@
+    // 2. sivu Alueet
+    override val alkuPvm: ZonedDateTime? = null,
+    override val loppuPvm: ZonedDateTime? = null,
+    override val tyoalueet: List<ApplicationArea>? = null,
+    // 3. sivu Yhteystiedot
+    override val yhteystiedot: Map<ApplicationContactType, HakemusyhteystietoRequest>? = null,
+    // 4. sivu Liitteet (separete endpoint)
+    // 5. sivu Yhteenveto (no input data)
+) : HakemusUpdateRequest {
+
+    override fun hasChanges(applicationEntity: ApplicationEntity): Boolean {
+        val applicationData = applicationEntity.applicationData as CableReportApplicationData
+        return nimi != applicationData.name ||
+            katuosoite != applicationData.postalAddress?.streetAddress?.streetName ||
+            rakennustyo != applicationData.constructionWork ||
+            huoltotyo != applicationData.maintenanceWork ||
+            kiinteistoliittyma != applicationData.propertyConnectivity ||
+            hatatyo != applicationData.emergencyWork ||
+            louhintaa != applicationData.rockExcavation ||
+            tyonKuvaus != applicationData.workDescription ||
+            alkuPvm != applicationData.startTime ||
+            loppuPvm != applicationData.endTime ||
+            tyoalueet != applicationData.areas ||
+            yhteystiedot.hasChanges(applicationEntity.yhteystiedot)
+    }
+}
+
+/**
+ * For updating an existing [Hakemusyhteystieto] (with [id]) or creating a new one (without [id]).
+ */
+data class HakemusyhteystietoRequest(
+    /** Hakemusyhteystieto id */
+    val id: UUID? = null,
+    val tyyppi: CustomerType? = null,
+    val nimi: String? = null,
+    val sahkoposti: String? = null,
+    val puhelin: String? = null,
+    val ytunnus: String? = null,
+    val yhteyshenkilot: List<HakemusyhteyshenkiloRequest>? = null,
+) {
+
+    /**
+     * Returns true if this customer has changes compared to the given [hakemusyhteystietoEntity].
+     */
+    fun hasChanges(hakemusyhteystietoEntity: HakemusyhteystietoEntity): Boolean =
+        tyyppi != hakemusyhteystietoEntity.tyyppi ||
+            nimi != hakemusyhteystietoEntity.nimi ||
+            sahkoposti != hakemusyhteystietoEntity.sahkoposti ||
+            puhelin != hakemusyhteystietoEntity.puhelinnumero ||
+            ytunnus != hakemusyhteystietoEntity.ytunnus ||
+            yhteyshenkilot.hasChanges(hakemusyhteystietoEntity.yhteyshenkilot)
+}
+
+data class HakemusyhteyshenkiloRequest(
+    /** Hankekayttaja id */
+    val id: UUID,
+    val tilaaja: Boolean
+)
+
+fun Map<ApplicationContactType, HakemusyhteystietoRequest>?.hasChanges(
+    hakemusyhteystiedot: Map<ApplicationContactType, HakemusyhteystietoEntity>
+): Boolean {
+    if (this == null) {
+        return hakemusyhteystiedot.isNotEmpty()
+    }
+    if (this.size != hakemusyhteystiedot.size) return true
+    return this.entries.any { (role, yhteystietoRequest) ->
+        hakemusyhteystiedot[role]?.let { entity -> yhteystietoRequest.hasChanges(entity) } ?: true
+    }
```

Comment:
If we check that the keys match instead of sizes, we know `hakemusyhteystiedot[role]` can't be null.

It's also easier to see why this covers cases where hakemusyhteystiedot has information this doesn't.
```suggestion
    if (this.keys != hakemusyhteystiedot.keys) return true
    return this.entries.any { (role, yhteystietoRequest) ->
        yhteystietoRequest.hasChanges(hakemusyhteystiedot[role]!!)
    }
```

Reply:
Thanks again! I use your suggestions.